src/lib.rs

```diff
@@ -1,506 +1,632 @@
+/// logger should use to keep the global maximum log level in sync with the
+/// highest log level that the logger will not ignore.
+///
+/// This function may only be called once in the lifetime of a program. Any log
+/// events that occur before the call to `set_logger` completes will be
+/// ignored.
+///
+/// This function does not typically need to be called manually. Logger
+/// implementations should provide an initialization method that calls
+/// `set_logger` internally.
+pub fn set_logger<M>(make_logger: M) -> Result<(), SetLoggerError>
+        where M: FnOnce(MaxLogLevelFilter) -> Box<Log> {
+    if LOGGER.compare_and_swap(UNINITIALIZED, INITIALIZING, Ordering::Relaxed) != UNINITIALIZED {
+        return Err(SetLoggerError);
+    }
 
-    unsafe {
-        LOG_LEVEL = max_level;
+    let logger = Box::new(make_logger(MaxLogLevelFilter(())));
+    let logger = unsafe { mem::transmute::<Box<Box<Log>>, usize>(logger) };
+    LOGGER.store(logger, Ordering::Release);
+    rt::at_exit(|| {
+        // Set to INITIALIZING to prevent re-initialization after
+        let logger = LOGGER.swap(INITIALIZING, Ordering::Acquire);
 
-        assert!(FILTER.is_null());
-        match filter {
-            Some(f) => FILTER = mem::transmute(Box::new(f)),
-            None => {}
+        while REFCOUNT.load(Ordering::Relaxed) != 0 {
+            // FIXME add a sleep here when it doesn't involve timers
         }
 
-        assert!(DIRECTIVES.is_null());
-        DIRECTIVES = mem::transmute(Box::new(directives));
-
-        // Schedule the cleanup for the globals for when the runtime exits.
-        rt::at_exit(|| {
-            assert!(!DIRECTIVES.is_null());
-            let _directives: Box<Vec<directive::LogDirective>> =
-                mem::transmute(DIRECTIVES);
-            DIRECTIVES = 0 as *const Vec<directive::LogDirective>;
-
-            if !FILTER.is_null() {
-                let _filter: Box<Regex> = mem::transmute(FILTER);
-                FILTER = 0 as *const _;
-            }
-        });
-    }
+        unsafe { mem::transmute::<usize, Box<Box<Log>>>(logger); }
+    });
+
+    Ok(())
 }
 
-#[cfg(test)]
-mod tests {
-    use super::enabled;
-    use directive::LogDirective;
-
-    #[test]
-    fn match_full_path() {
-        let dirs = [
-            LogDirective {
-                name: Some("crate2".to_string()),
-                level: 3
-            },
-            LogDirective {
-                name: Some("crate1::mod1".to_string()),
-                level: 2
-            }
-        ];
-        assert!(enabled(2, "crate1::mod1", dirs.iter()));
-        assert!(!enabled(3, "crate1::mod1", dirs.iter()));
-        assert!(enabled(3, "crate2", dirs.iter()));
-        assert!(!enabled(4, "crate2", dirs.iter()));
+/// The type returned by `set_logger` if `set_logger` has already been called.
+#[allow(missing_copy_implementations)]
+#[derive(Debug)]
+pub struct SetLoggerError;
```

Comment:
This may want to have a private field to prevent construction (in case it gets expanded later)